render-ws-java-client/src/main/java/org/janelia/render/client/match/UnconnectedTileEdges.java

```diff
@@ -0,0 +1,215 @@
+        for (final CellEdge edge : sortedEdges) {
+            for (final ZRange zRange : edgeToUnconnectedZRangesMap.get(edge)) {
+                sortedUnconnectedEdges.add("edge " + edge + " is not connected for " + zRange);
+            }
+        }
+        return sortedUnconnectedEdges;
+    }
+
+    public boolean hasTooManyConsecutiveUnconnectedEdges() {
+        return edgeToUnconnectedZRangesMap.values().stream()
+                .flatMap(Collection::stream)
+                .anyMatch(zRange -> zRange.size() > maxUnconnectedLayers);
+    }
+
+    private List<CellEdge> buildExpectedConnectedEdges(final List<Integer> sortedRows,
+                                                       final List<Integer> sortedColumns,
+                                                       final Set<CellId> cellsForZ) {
+        final List<CellEdge> expectedConnectedEdges = new ArrayList<>();
+
+        for (final Integer row : sortedRows) {
+            for (int i = 1; i < sortedColumns.size(); i++) {
+                final CellId previousCell = new CellId(row, sortedColumns.get(i - 1));
+                final CellId cell = new CellId(row, sortedColumns.get(i));
+
+                if (((cell.column - previousCell.column) == 1) &&
+                    cellsForZ.contains(previousCell) &&
+                    cellsForZ.contains(cell)) {
+
+                    expectedConnectedEdges.add(new CellEdge(previousCell, cell));
+                } // else gap in columns so skip
+            }
+        }
+
+        for (final Integer column : sortedColumns) {
+            for (int i = 1; i < sortedRows.size(); i++) {
+                final CellId previousCell = new CellId(sortedRows.get(i - 1), column);
+                final CellId cell = new CellId(sortedRows.get(i), column);
+
+                if ((cell.row - previousCell.row) == 1 &&
+                    cellsForZ.contains(previousCell) &&
+                    cellsForZ.contains(cell)) {
+
+                    expectedConnectedEdges.add(new CellEdge(previousCell, cell));
+                } // else gap in rows so skip
+            }
+        }
+
+        return expectedConnectedEdges;
+    }
+
+    @Nonnull
+    private Set<CellEdge> buildActualConnectedEdges(final ResolvedTileSpecCollection resolvedTiles,
+                                                    final List<CanvasMatches> matchedPairs) {
+        final Set<CellEdge> connectedEdges = new HashSet<>();
+        for (final CanvasMatches canvasMatches : matchedPairs) {
+            // only consider matches within the same group
+            if (canvasMatches.getpGroupId().equals(canvasMatches.getqGroupId())) {
+                try {
+                    connectedEdges.add(CellEdge.getEdgeForPair(canvasMatches.getpId(),
+                                                               canvasMatches.getqId(),
+                                                               resolvedTiles));
+                } catch (final IllegalArgumentException e) {
```

Comment:
Nice catch!  Although putting a null entry in the HashSet would not cause an exception, it would quietly hide that the stack has tile specs missing required layout data.  I updated the code to log this.